[#5581] implement 1p vs 3p Alfen charger detection

ikruglov opened the pull request:
Fix https://github.com/evcc-io/evcc/issues/5581

This PR implement auto-detection of 1p or 3p Alfen charger.

The autodetection logic works by looking into phase voltages. If it sees voltage on L1 but no voltage on L2 and L3 it assumes that the charger is 1p. In all other cases, the logic defaults to 3p for backward compatibility.

See problem and the description in https://github.com/evcc-io/evcc/issues/5581

ikruglov committed:
[#5581] implement 1p vs 3p Alfen charger detection

andig commented:
Could you rebase on master? Then remove the 1p template, make docs and take it from there. We need the decorator anyway!

andig commented:
Also, lets add a Voltages function, we‘ll need that later for another PR anyway.

ikruglov commented:
> Could you rebase on master? Then remove the 1p template, make docs and take it from there. We need the decorator anyway!
Will do

> lets add a Voltages function, we‘ll need that later for another PR anyway.
There is one already!

andig commented:
And of course remove the phases setting again, this is replaced by your detection.

andig commented:
> There is one already

an exported function with the actual name, signature like currents

ikruglov commented:
> an exported function with the actual name, signature like currents

I already implemented Voltage function as you're describing, no?

ikruglov committed:
Revert "Alfen: add 1p template"

This reverts commit 70454222b01e9375463cfc75df32385ffcef1a6f.

andig commented:
Ah, didnt see it on mobile, perfect

ikruglov commented:
@andig, I've pulled, reverted your commit, and pushed my on top. Pushed with force to my branch after that.

andig commented on charger/alfen.go:
```diff
@@ -244,9 +251,22 @@ func (wb *Alfen) Currents() (float64, float64, float64, error) {
-// phases1p3p implements the api.PhaseSwitcher interface
-func (c *Alfen) phases1p3p(phases int) error {
-	_, err := c.conn.WriteSingleRegister(alfenRegPhases, uint16(phases))
+func (wb *Alfen) is3p() bool {
```
This is a bit complex. Lets just check in the New function? Err must be nil and phases match value/nan/nan?

ikruglov replied:
This is exactly what the function currently does. Do you want to remove logging and shorten the code?

andig commented on charger/alfen.go:
```diff
@@ -244,9 +251,22 @@ func (wb *Alfen) Currents() (float64, float64, float64, error) {
+		return true
+	}
+
+	return false // defaults to 3p charger
```
This is really is1p ;)

ikruglov replied:
ah, right :-)

ikruglov replied:
renames to is1p

ikruglov committed:
add docs

ikruglov committed:
rename is3p to is1p and fix a tiny bug

ikruglov commented:
Added docs, fixed is3p -> is1p, and awaiting your comment on moving `is1p()` inside `New()`. Logically it's a noop but the explicit `is1p()` function lets me have more verbose output. Which I prefer for ease of debugging. But not a strong opinion.

ikruglov committed:
fix typo in charger/openevse.go which I reverted

andig committed:
Simplify detection

andig commented:
Pushed 1 commit to simplify

ikruglov commented:
please wait a bit and do not merge. I'm getting 3p detection on my charge with this code. Something is wrong.

andig committed:
fix current

andig commented:
Spotted it- the NaNs are already removed...

ikruglov commented:
yeah, `voltagesOrCurrents` converts IsNaN to 0 :-(. Do you want to check based on 0s, or better to expose raw values?

ikruglov commented:
I'd prefer to do raw values. Don't like magic conversions

ikruglov committed:
use zeros instead of NaN to detect voltage

ikruglov commented:
However, maybe I'm too paranoid. Pushed based on 0. Let me know what you think.

ikruglov committed:
keep verbosity for debugging

andig commented:
Probably just as good- lets merge?

ikruglov commented:
yup, tested and it works. Let's merge it

andig commented:
Thank you, much appreciated!

ikruglov commented:
@andig you are welcome and thank you too!

On a side note. I've actually come to evcc for the load-balancing feature which is currently missing. I've ready it somewhere that it's on the todo list. Is there any conversation about it? I'm willing to implement or contribute to implementing it depending on its complexity.

andig commented:
Sure, please see https://github.com/evcc-io/evcc/pull/3453 or join us on Slack!
